src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -168,6 +219,221 @@ private static IDataTransform Create(IHostEnvironment env, Arguments args, IData
+            env.CheckValue(args, nameof(args));
+            env.CheckValue(input, nameof(input));
+            env.CheckNonEmpty(args.OptimizationOperation, nameof(args.OptimizationOperation));
+
+            if (TensorFlowUtils.IsSavedModel(env, model))
+                TrainCore(args, model, input);
+            else
+                throw env.ExceptNotSupp("TensorFlowTransform: Re-Training of TensorFlow model is only possible for un-frozen model.");
+        }
+
+        private void TrainCore(TrainingArguments args, string model, IDataView input)
+        {
+            var inputsForTraining = new string[Inputs.Length + 1];
+            var inputColIndices = new int[inputsForTraining.Length];
+            var isInputVector = new bool[inputsForTraining.Length];
+            var tfInputTypes = new TFDataType[inputsForTraining.Length];
+            var tfInputShapes = new TFShape[inputsForTraining.Length];
+
+            for (int i = 0; i < Inputs.Length; i++)
+            {
+                inputsForTraining[i] = Inputs[i];
+            }
+
+            inputsForTraining[inputsForTraining.Length - 1] = args.LabeLColumn;
+
+            var inputSchema = input.Schema;
+            for (int i = 0; i < inputsForTraining.Length; i++)
+            {
+                if (!inputSchema.TryGetColumnIndex(inputsForTraining[i], out inputColIndices[i]))
+                    throw _host.Except($"Column {inputsForTraining[i]} doesn't exist");
+
+                var type = inputSchema.GetColumnType(inputColIndices[i]);
+                isInputVector[i] = type.IsVector;
+
+                var tfInput = new TFOutput(Graph[inputsForTraining[i]]);
+                tfInputTypes[i] = tfInput.OutputType;
+                tfInputShapes[i] = Graph.GetTensorShape(tfInput);
+                if (tfInputShapes[i].NumDimensions != -1)
+                {
+                    var newShape = new long[tfInputShapes[i].NumDimensions];
+                    newShape[0] = tfInputShapes[i][0] == -1 ? args.BatchSize : tfInputShapes[i][0];
+
+                    for (int j = 1; j < tfInputShapes[i].NumDimensions; j++)
+                        newShape[j] = tfInputShapes[i][j];
+                    tfInputShapes[i] = new TFShape(newShape);
+                }
+
+                var expectedType = TensorFlowUtils.Tf2MlNetType(tfInputTypes[i]);
+                if (type.ItemType != expectedType)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", inputsForTraining[i], expectedType.ToString(), type.ToString());
+            }
+
+            var fetchList = new List<string>();
+            if (args.LossOperation != null)
+                fetchList.Add(args.LossOperation);
+            if (args.MetricOperation != null)
+                fetchList.Add(args.MetricOperation);
+
+            for (int epoch = 0; epoch < args.Epoch; epoch++)
+            {
+                using (var cursor = input.GetRowCursor(a => true))
```

Comment:
>a => true [](start = 55, length = 9)

a=>  (a) in inputColIndices
(you can create hashSet and just call hashSet.Contains function)
But in general, it doesn't make much sense to ask for all columns, and consume only few of them.

Reply:
technically you can just pass hashedIndices.Contains as predicate, all this marshalling of variable a, is not necessary :)

---
In reply to: [221766659](https://github.com/dotnet/machinelearning/pull/1063#discussion_r221766659) [](ancestors = 221766659)